File: problem206.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if(head == nullptr) return head;
        ListNode* tail;
        ListNode* dummy = head;
        head = head->next;
        dummy->next = nullptr;
        while(head){
            tail = dummy;
            dummy = head;
            head = head->next;
            dummy->next = tail;
        }
        return dummy;
    }
};

File: Grind75/Week 2/problem206.py
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        tail = ListNode(head.val)
        head = head.next
        while head.next:
            temp = head
            tempNext = head.next
            head.next = tail
            tail = temp
            head = tempNext
        head.next = tail
        return head